Reset CI tests when PR tags are set back to 'Ready'

TerrenceMcGuinness-NOAA opened the pull request:
**Description**

**bug** fix **feature** that adds updates to the CI cron BASH scripts that track an ID based on how many times the label gets updated to Ready.  This ID count is used in the driver script so that it can distinguish a real fail and one caused by removing the current build that is overwritten by a new Ready request.

Fixes #1649

**Type of change**

updates to drive scripts so a CI run can stop and restart midstream.

- [x] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

**How Has This Been Tested?**

I updated the CI label to **Ready** when it was **Building** and it did not create a false fail.

**Checklist**

- [x] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

TerrenceMcGuinness-NOAA committed:
added code to the driver script to reset CI when label is switched back to Ready once in progress

TerrenceMcGuinness-NOAA committed:
missed second star for wildcard for substring test in bash for when driver get Ready when had already been

TerrenceMcGuinness-NOAA committed:
had substring logic in reverse for driver script

TerrenceMcGuinness-NOAA committed:
added a Ready update count as an id so that multiple driver instances will not create false negatives on build when starting over in PR dir

TerrenceMcGuinness-NOAA committed:
pynorms - few basic ones

aerorahul commented on ci/scripts/check_ci.sh:
```diff
@@ -57,7 +57,7 @@ pr_list_dbfile="${GFS_CI_ROOT}/open_pr_list.db"
 pr_list=""
 if [[ -f "${pr_list_dbfile}" ]]; then
-  pr_list=$("${HOMEgfs}/ci/scripts/pr_list_database.py" --display "${pr_list_dbfile}" | grep -v Failed | grep Running | awk '{print $1}') || true
+  pr_list=$("${HOMEgfs}/ci/scripts/pr_list_database.py" --display --sbfild "${pr_list_dbfile}" | grep -v Failed | grep Running | awk '{print $1}') || true
```
Should this be `--sbfile`?  Is this a `typo`?

TerrenceMcGuinness-NOAA replied:
yes it should, updating.

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -59,18 +59,38 @@ set -x
+  pr_id=0
+#############################################################
+# Check if a Ready labeled PR has changed back from once set 
+# and in that case remove all previous jobs in slubm and
```
```suggestion
# and in that case remove all previous jobs in scheduler and
```

aerorahul commented on ci/scripts/driver.sh:
```diff
@@ -59,18 +59,38 @@ set -x
+         break
+      fi   
+      pslot=$(basename "${cases}")
+      sacct --format=jobid,jobname%35,WorkDir%100,stat | grep "${pslot}" | grep "PR\/${pr}\/RUNTESTS" |  awk '{print $1}' | xargs scancel || true
```
I see you are hard-wiring `sacct` for `slurm`.  Fine for now, but beware when you try and do this on WCOSS2 which has `PBS`.

TerrenceMcGuinness-NOAA replied:
Oh yes, good to know!

aerorahul commented on ci/scripts/pr_list_database.py:
```diff
@@ -2,6 +2,7 @@
 
 import sys
 from pathlib import Path
+import argparse
```
Aren't you already importing what you need from `argparse` on line 6?  What is this for?

TerrenceMcGuinness-NOAA replied:
actually, I needed it for `argparse.REMAINDER` on line 126

WalterKolczynski-NOAA replied:
Why not just import that then?

TerrenceMcGuinness-NOAA replied:
ok

aerorahul reviewed (commented):
Looks mostly ok.
A typo was spotted, and a clarification is sought.

TerrenceMcGuinness-NOAA committed:
needed to add --sbfile switch onto pr_list_database.py script in run_ci.sh

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/driver.sh

Co-authored-by: Rahul Mahajan <aerorahul@users.noreply.github.com>

TerrenceMcGuinness-NOAA committed:
Update check_ci.sh

Type spelling file for new switch to pr_list_database.py in check

TerrenceMcGuinness-NOAA reviewed (commented):
@aerorahul  : updated typo and responded to comments and questions

WalterKolczynski-NOAA commented on ci/scripts/driver.sh:
```diff
@@ -59,18 +59,38 @@ set -x
+# Check if a Ready labeled PR has changed back from once set 
+# and in that case remove all previous jobs in scheduler and
+# and remove PR from filesystem to start clean
+#############################################################
```
Comment indentation should match the block it is contained in.

WalterKolczynski-NOAA commented on ci/scripts/pr_list_database.py:
```diff
@@ -51,14 +52,14 @@ def sql_insert(obj: sqlite3.Cursor, entities: list) -> None:
         Cursor object for Sqlite3
     entities : list
-        The list three string values that go into sqlite table (pr, state, status)
+        The list three string values that go into sqlite table (pr, state, status, reset)
```
```suggestion
        A list of four string values that go into sqlite table (pr, state, status, reset)
            pr: pull request number
            state: The new value for the state (Open, Closed)
            status: The new value for the status (Ready, Running, Failed)
            reset: The value for number of times reset to Ready
```

WalterKolczynski-NOAA commented on ci/scripts/pr_list_database.py:
```diff
@@ -114,11 +119,12 @@ def input_args():
                             formatter_class=ArgumentDefaultsHelpFormatter)
 
-    parser.add_argument('sbfile', help='SQLite3 database file with PR list', type=str)
+    parser.add_argument('--sbfile', help='SQLite3 database file with PR list')
```
Why was the type removed?

TerrenceMcGuinness-NOAA replied:
"type removed"?? I added this argument as a switch instead it being positional because the update argument list was no longer fixed.

TerrenceMcGuinness-NOAA replied:
Oh wait, I missed the wrap around, yes the type was changed because it is being passed as a list now.  You can see the list is applied to the database in a loop depending the number of arguments.  This may need work and we have decided to improve on this interface for another PR.

WalterKolczynski-NOAA replied:
I just went through the PR and I do not see anywhere where `--sbfile` is passed multiple files. Further, I do not see a use case where we would want to use multiple _databases_.

Also, if you wanted to have a list, shouldn't you would use `nargs`?:
```suggestion
    parser.add_argument('--sbfiles', help='SQLite3 database file with PR list', nargs='+', type=str)
```

TerrenceMcGuinness-NOAA replied:
Oh, I misspoke,  `sbfiles` is no longer positional because `update_pr` takes a variable number of arguments and there the type change here was for another reason.

TerrenceMcGuinness-NOAA replied:
BTW, the type was removed, not changed, because it had to take on a path and there was an issue with it being a string, not sure why that matters.

WalterKolczynski-NOAA replied:
I'm not sure why `str` wouldn't work anymore, but I'll take your word for it. Leave this unresolved in case Rahul wants to chime in.

TerrenceMcGuinness-NOAA replied:
me neither, let me check ...

`(gfs_workflow) mterry (Orion-login-1) scripts (bug_1649_Ready-Restart) $ git diff pr_list_database.py 
diff --git a/ci/scripts/pr_list_database.py b/ci/scripts/pr_list_database.py
index 8f519e0f..a099ff51 100755
--- a/ci/scripts/pr_list_database.py
+++ b/ci/scripts/pr_list_database.py
@@ -122,7 +122,7 @@ def input_args():
     parser = ArgumentParser(description=description,
                             formatter_class=ArgumentDefaultsHelpFormatter)
 
-    parser.add_argument('--sbfile', help='SQLite3 database file with PR list')
+    parser.add_argument('--sbfile', help='SQLite3 database file with PR list', type=str)
     parser.add_argument('--create', help='create sqlite file for pr list status', action='store_true', required=False)
     parser.add_argument('--add_pr', nargs=1, metavar='PR', help='add new pr to list (defults to: Open,Ready)', required=False)
     parser.add_argument('--remove_pr', nargs=1, metavar='PR', help='removes pr from list', required=False)
(gfs_workflow) mterry (Orion-login-1) scripts (bug_1649_Ready-Restart) $ ./pr_list_database.py --display --sbfile /work2/noaa/stmp/GFS_CI_ROOT/open_pr_list.db
1601 Open Built 0`

Yep, no big deal here.

TerrenceMcGuinness-NOAA replied:
Me neither, here let me check ...

TerrenceMcGuinness-NOAA replied:
```
(gfs_workflow) mterry (Orion-login-1) scripts (bug_1649_Ready-Restart) $ git diff pr_list_database.py 
diff --git a/ci/scripts/pr_list_database.py b/ci/scripts/pr_list_database.py
index 8f519e0f..a099ff51 100755
--- a/ci/scripts/pr_list_database.py
+++ b/ci/scripts/pr_list_database.py
@@ -122,7 +122,7 @@ def input_args():
     parser = ArgumentParser(description=description,
                             formatter_class=ArgumentDefaultsHelpFormatter)
 
-    parser.add_argument('--sbfile', help='SQLite3 database file with PR list')
+    parser.add_argument('--sbfile', help='SQLite3 database file with PR list', type=str)
     parser.add_argument('--create', help='create sqlite file for pr list status', action='store_true', required=False)
     parser.add_argument('--add_pr', nargs=1, metavar='PR', help='add new pr to list (defults to: Open,Ready)', required=False)
     parser.add_argument('--remove_pr', nargs=1, metavar='PR', help='removes pr from list', required=False)
(gfs_workflow) mterry (Orion-login-1) scripts (bug_1649_Ready-Restart) $ ./pr_list_database.py --display --sbfile /work2/noaa/stmp/GFS_CI_ROOT/open_pr_list.db
1601 Open Built 0
```

TerrenceMcGuinness-NOAA replied:
Yep, no big deal.

WalterKolczynski-NOAA commented on ci/scripts/driver.sh:
```diff
@@ -85,23 +105,35 @@ fi
   set +e
   "${HOMEgfs}/ci/scripts/clone-build_ci.sh" -p "${pr}" -d "${pr_dir}" -o "${pr_dir}/output_${id}"
   ci_status=$?
+##################################################################
+# Checking for special case when Ready label was updated
+# that cause a running driver exit fail because was currently
+# building so we force and exit 0 instead to does not get relabled
+#################################################################
```
Add indentation to match surroundings.

TerrenceMcGuinness-NOAA committed:
Update ci/scripts/pr_list_database.py

Co-authored-by: Walter Kolczynski - NOAA <Walter.Kolczynski@noaa.gov>

TerrenceMcGuinness-NOAA committed:
Update driver.sh

spacing

TerrenceMcGuinness-NOAA reviewed (commented):
spacing - import

TerrenceMcGuinness-NOAA committed:
Merge branch 'NOAA-EMC:develop' into bug_1649_Ready-Restart

TerrenceMcGuinness-NOAA committed:
import const in argparse better

TerrenceMcGuinness-NOAA reviewed (commented):
added comment after understanding the question better from Walter on the arg list for the database update script

TerrenceMcGuinness-NOAA committed:
Update pr_list_database.py

trailing space

TerrenceMcGuinness-NOAA reviewed (commented):
submit review

TerrenceMcGuinness-NOAA committed:
Update pr_list_database.py

put type=str back in for sbfile switch in argparser for the control database python script for the CI state manager

TerrenceMcGuinness-NOAA reviewed (commented):
re-submit review
